nexus/src/nexus.rs

```diff
@@ -1173,23 +1174,33 @@ impl Nexus {
         &self,
         instance_id: Uuid,
         migration_id: Uuid,
+        dst_propolis_id: Uuid,
     ) -> UpdateResult<db::model::Instance> {
         let instance = self.datastore().instance_fetch(&instance_id).await?;
 
         let requested = InstanceRuntimeStateRequested {
             run_state: InstanceStateRequested::Migrating,
-            migration_id: Some(migration_id),
+            migration_params: Some(InstanceRuntimeStateMigrateParams {
+                migration_id,
+                dst_propolis_id,
+            }),
         };
         self.check_runtime_change_allowed(
             &instance.runtime().clone().into(),
             &requested,
         )?;
-        self.instance_set_runtime(
-            &instance,
-            self.instance_sled(&instance).await?,
-            requested,
-        )
-        .await?;
+        let updated = self
+            .instance_set_runtime(
+                &instance,
+                self.instance_sled(&instance).await?,
+                requested,
+            )
+            .await?;
+        if !updated {
+            return Err(Error::internal_error(
+                "couldn't update instance record for migrate",
+            ));
+        }
```

Comment:
So, as a heads-up: Under the hood, the datastore's `instance_update_runtime` uses the `update_and_check` CTE to basically try to perform a few operations simultaneously:

- Check if the instance exists, with all the associated `WHERE` clauses (time deleted, ID = instance ID, generation number < the expected one), and update the runtime if those conditions are true
- If *any* of them aren't true, identify if the instance exists. `false` can be returned if the instance exists but has been deleted, the generation number is too low, etc.

This case, where we return `updated = false`, can happen when "the instance exists but the Sled Agent sent us a more recent response". Perhaps the state transitions in the Sled Agent happened like this:

- Sled Agent: Gen = 1: Instance Running
- (Migrate operation) Nexus sends "migration" request to SA
- Sled Agent: Gen = 2: Instance Migrating
- (Migrate operation) Nexus receives response: `new_runtime = Gen 2`.
- (Concurrent operation) Nexus sends "Go to new state" request to SA
- Sled Agent: Gen = 3: Instance Enters New State
- (Concurrent operation) Nexus receives response: `new_runtime = Gen 3`
- (Concurrent operation) Nexus updates DB with `runtime` for `Gen = 3`
- (Migrate operation) Nexus updates DB with `runtime` for `Gen = 2`

In this case, `instance_update_runtime` would return `false`, we'd hit this case, and return a 500 error.
**However**, as long as this info was sent back to Nexus from the Sled Agent - and correctly updated then - there's no reason for us to return 500, right?

Since this case seems possible with a concurrent operation, should we be propagating it back to users? A 500 to me is basically a sign that "the server is broken, or we have a bug", but it seems like we have control over this case.

Reply:
That's exactly the scenario this was trying to catch: we want to be make sure we're actually in the Migrating state before proceeding onto the next migration saga action. Alternatively, is there a way to indicate a saga action should be re-run until it successfully returns?